core/src/main/scala/org/apache/spark/sql/delta/files/TahoeRemoveFileIndex.scala

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright (2021) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.delta.files
+
+import org.apache.spark.sql.delta.{DeltaLog, Snapshot}
+import org.apache.spark.sql.delta.actions.{AddFile, RemoveFile}
+import org.apache.spark.sql.delta.actions.SingleAction.addFileEncoder
+import org.apache.spark.sql.delta.commands.cdc.CDCReader
+import org.apache.spark.sql.delta.commands.cdc.CDCReader._
+import org.apache.hadoop.fs.Path
+
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.catalyst.expressions.Expression
+import org.apache.spark.sql.types.StructType
+
+/**
+ * A [[TahoeFileIndex]] for scanning a sequence of removed files as CDC. Similar to
+ * [[TahoeBatchFileIndex]], the equivalent for reading [[AddFile]] actions.
+ */
+class TahoeRemoveFileIndex(
+    spark: SparkSession,
+    val filesByVersion: Seq[CDCDataSpec[RemoveFile]],
+    deltaLog: DeltaLog,
+    path: Path,
+    snapshot: Snapshot) extends TahoeFileIndex(spark, deltaLog, path) {
+
+  override def tableVersion: Long = snapshot.version
```

Comment:
I see a lot of the code between TahoeAddFileIndex and TahoeRemoveFileIndex as common. couldnt there have been an abstract class to dedup the code?

Reply:
I see. In the interest of unblocking your follow up PRs (which I can see from the design, there are going to be a few), I am okay if you refactor this later.